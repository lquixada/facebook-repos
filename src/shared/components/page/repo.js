import React from 'react'
import {Helmet} from 'react-helmet'
import styled from 'styled-components'

import Repo from '../../containers/repo'
import Menu from '../../containers/menu'
import {Footer} from '../footer'
import {Aside, Content, Section, SectionContainer} from '../grid'
import {flex} from '../../helpers'

export const RepoPage = ({repoName}) => (
  <Section>
    <Helmet>
      <title>{repoName ? `${repoName} -` : ''}Github Repos</title>
    </Helmet>

    <SectionContainer>
      <Aside>
        <Menu />
        <Footer />
      </Aside>
      <Content>
        {repoName
          ? <Repo />
          : (
            <P>
              Choose a repository on the menu.
            </P>
          )
        }
      </Content>
    </SectionContainer>
  </Section>
)

const P = styled.p`
  ${flex.display()}
  ${flex.middle()} 
  ${flex.center()}
  height: 50rem;
  border: .7rem dashed #eee;
  border-radius: .6rem;
  font-size: 1.4rem;
  letter-spacing: -.1px;
`
