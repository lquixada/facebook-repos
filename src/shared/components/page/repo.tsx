import React from 'react'
import {Helmet} from 'react-helmet'
import styled from 'styled-components'

import Menu from '../../containers/menu'
import Repo from '../../containers/repo'
import {flex} from '../../helpers'
import {Footer} from '../footer'
import {Aside, Content, Section, SectionContainer} from '../grid'

interface IProps {
  repoName: string | undefined
}

export const RepoPage: React.SFC<IProps> = ({repoName}) => (
  <Section>
    <Helmet>
      <title>{repoName ? `${repoName} -` : ''}Github Repos</title>
    </Helmet>

    <SectionContainer>
      <Aside>
        <Menu />
        <Footer />
      </Aside>
      <Content>
        {repoName
          ? <Repo />
          : (
            <P>
              Choose a repository on the menu.
            </P>
          )
        }
      </Content>
    </SectionContainer>
  </Section>
) as React.ReactElement<any>

const P = styled.p`
  ${flex.display()}
  ${flex.middle()}
  ${flex.center()}
  height: 50rem;
  border: .7rem dashed #eee;
  border-radius: .6rem;
  font-size: 1.4rem;
  letter-spacing: -.1px;
`
