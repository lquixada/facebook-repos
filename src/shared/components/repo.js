import React from 'react';
import styled from 'styled-components';

export class Repo extends React.Component {
  getRepo() {
    const {repos, match} = this.props;
    return repos.find((repo) => repo.name === match.params.repo);
  }

  render() {
    return (
      <Wrapper>
        <pre>
          {JSON.stringify(this.getRepo(), null, '\t')}
        </pre>
      </Wrapper>
    );
  }
}

const Wrapper = styled.div`
  color: red;
`;
