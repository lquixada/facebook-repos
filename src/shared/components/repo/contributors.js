import React from 'react';
import styled from 'styled-components';

import {flex} from '../../helpers';

export class Contributors extends React.Component {
  constructor(props) {
    super(props);
    this.fetchNext = this.fetchNext.bind(this);
  }

  componentDidMount() {
    this.fetch();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.repoName !== this.props.repoName) {
      this.fetch();
    }
  }

  fetch() {
    if (!this.hasContributorsLoaded()) {
      this.props.fetchContributors(this.props.repoName);
    }
  }

  fetchNext() {
    const {repoName, contributors} = this.props;
    this.props.fetchMoreContributors(repoName, contributors.next);
  }

  hasContributorsLoaded() {
    const {contributors} = this.props;
    return contributors && contributors.result && contributors.result.length !== 0;
  }

  hasMoreContributors() {
    return !!this.props.contributors.next;
  }

  isLoading() {
    return this.props.contributors.isLoading;
  }

  getCount() {
    const {count} = this.props;
    return count? count[1] : 'error';
  }

  render() {
    return (
      <Wrapper>
        <Title>Contributors ({this.getCount()})</Title>
        {this.renderContent()}
      </Wrapper>
    );
  }

  renderContent() {
    if (!this.hasContributorsLoaded()) {
      return (
        <div>Loading...</div>
      );
    }

    return (
      <div>
        <List>
          {this.props.contributors.result.map((contributor, i) =>
            <Item key={contributor.login+i}>
              <Link href={contributor.html_url}>
                <Image src={contributor.avatar_url} />
                <Text>{contributor.login}</Text>
              </Link>
            </Item>
          )}
        </List>
        {this.hasMoreContributors() &&
          <Button onClick={this.fetchNext}>
            {this.isLoading()? 'Loading...' : 'More'}
          </Button>
        }
      </div>
    );
  }
}

const Wrapper = styled.div`
  margin: 2rem 0;
`;

const Title = styled.h3`
  color: #767676;
  font-size: 1.4rem;
  font-weight: normal;
  line-height: 1.5;
`;

const List = styled.ul`
  ${flex.display()}
  ${flex.left()}
  ${flex.flow('row', 'wrap')};
  margin-top: 1rem;
`;

const Item = styled.li`
  ${flex.display()}
  ${flex.left()}
  ${flex.middle()}
  margin-bottom: 1rem;
  width: 25%;
  font-size: 1.2rem;
  white-space: nowrap;
`;

const Link = styled.a`
  ${flex.display()}
  ${flex.left()}
  ${flex.middle()}
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
`;

const Image = styled.img`
  margin-right: 1rem;
  width: 4rem;
  height: 4rem;
  border-radius: 2rem;
  overflow: hidden;
`;

const Text = styled.span`
  display: inline-block;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
`;

const Button = styled.button`
  display: block;
  width: 100%;
  padding: .6rem;
  margin-top: 2rem;
  font-size: 1.4rem;
  font-weight: 600;
  color: #0366d6;
  background: #fff;
  border: 1px solid #e1e4e8;
  border-radius: .3rem;
  cursor: pointer;

  &:hover, &:focus {
    background-color: #f6f8fa;
  }
`;
