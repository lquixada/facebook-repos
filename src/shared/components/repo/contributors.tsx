import React from 'react'
import {NavLink} from 'react-router-dom'
import styled from 'styled-components'

import {flex, tablets} from '../../helpers'

interface IContributor {
  login: string
  avatar_url: string
  html_url: string
}
interface IProps {
  data: IContributor[]
  total: string | number
  hasLoaded: boolean
  hasMore: boolean
  isLoadingMore: boolean
  onNext(f: any): any
}

export const Contributors = (props: IProps) => (
  <Wrapper>
    <Title>Contributors ({props.total})</Title>
    {!props.hasLoaded ? <Loading /> : <Content {...props} />}
  </Wrapper>
)

const Loading = () => (
  <div>Loading...</div>
)

const Content = ({data, hasMore, isLoadingMore, onNext}: IProps) => (
  <div>
    <List>
      {data.map((contributor, i) =>
        <Item key={contributor.login + i}>
          <Link to={`/r/${contributor.login}`}>
            <Image alt={`User ${contributor.login}`} src={`${contributor.avatar_url}&s=40`} />
            <Text>{contributor.login}</Text>
          </Link>
        </Item>,
      )}
    </List>
    {hasMore &&
      <Button onClick={onNext}>
        {isLoadingMore ? 'Loading...' : 'More'}
      </Button>
    }
  </div>
)

const Wrapper = styled.div`
  margin: 2rem 0;
`

const Title = styled.h3`
  color: #767676;
  font-size: 1.4rem;
  font-weight: normal;
  line-height: 1.5;
`

const List = styled.ul`
  ${flex.display()}
  ${flex.left()}
  ${flex.flow('row', 'wrap')};
  margin-top: 1rem;
`

const Item = styled.li`
  ${flex.display()}
  ${flex.left()}
  ${flex.middle()}
  margin-bottom: 1rem;
  width: 25%;
  font-size: 1.2rem;
  white-space: nowrap;

  @media (${tablets}) {
    width: 33.33%;
  }
`

const Link = styled(NavLink)`
  ${flex.display()}
  ${flex.left()}
  ${flex.middle()}
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
`

const Image = styled.img`
  margin-right: 1rem;
  width: 4rem;
  height: 4rem;
  border-radius: 2rem;
  overflow: hidden;
`

const Text = styled.span`
  display: inline-block;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
`

const Button = styled.button`
  display: block;
  width: 100%;
  padding: .6rem;
  margin-top: 2rem;
  font-size: 1.4rem;
  font-weight: 600;
  color: #0366d6;
  background: #fff;
  border: 1px solid #e1e4e8;
  border-radius: .3rem;
  cursor: pointer;

  &:hover, &:focus {
    background-color: #f6f8fa;
  }
`
