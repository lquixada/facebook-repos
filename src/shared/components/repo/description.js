import React from 'react';
import styled from 'styled-components';

export const Description = ({text}) => (
  <P>{text}</P>
);

const P = styled.p`
  color: #393939;
  font-size: 1.4rem;
  line-height: 1.5;
`;
