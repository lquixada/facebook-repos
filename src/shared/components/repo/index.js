import React from 'react';

import Contributors from '../../containers/repo/contributors';
import {Description} from './description';
import {Header} from './header';
import {Summary} from './summary';

export const Repo = ({repo}) => (
  <div>
    <Header title={repo.get('name')} url={repo.get('html_url')} />
    <Description text={repo.get('description')} />
    <Summary repo={repo} />
    <Contributors repoName={repo.get('name')} />
  </div>
);
