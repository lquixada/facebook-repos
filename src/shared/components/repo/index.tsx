import {Map} from 'immutable'
import React from 'react'

import Contributors from '../../containers/repo/contributors'
import {Description} from './description'
import {Header} from './header'
import {Summary} from './summary'

interface IProps {
  owner: string
  repo: object
}

export const Repo = ({owner, repo}: IProps) => (
  <div>
    <Header title={repo.name} url={repo.html_url} />
    <Description text={repo.description} />
    <Summary owner={owner} repo={repo} />
    <Contributors owner={owner} repoName={repo.name} />
  </div>
)
