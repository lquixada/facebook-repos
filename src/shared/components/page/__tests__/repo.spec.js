import React from 'react'
import {shallow} from 'enzyme'

import {RepoPage} from '../repo'

describe('<RepoPage />', () => {
  it('renders the repo info when repo name is given', () => {
    const component = shallow(<RepoPage repoName='repo1' />)
    expect(component).toMatchSnapshot()
  })

  it('renders a selection message when repo name is not given', () => {
    const component = shallow(<RepoPage />)
    expect(component).toMatchSnapshot()
  })
})
