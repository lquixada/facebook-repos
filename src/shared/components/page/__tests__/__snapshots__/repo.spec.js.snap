// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RepoPage /> matches the snapshot 1`] = `
<styled.section>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
       - Github Repos
    </title>
  </HelmetWrapper>
  <styled.div>
    <styled.aside>
      <withRouter(Connect(MenuContainer)) />
      <Footer />
    </styled.aside>
    <styled.article>
      <withRouter(Connect(RepoContainer)) />
    </styled.article>
  </styled.div>
</styled.section>
`;
